## 練習時間
試著使用 sklearn datasets 的其他資料集 (boston, ...)，來訓練自己的線性迴歸模型，並加上適當的正則話來觀察訓練情形。

In [6]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn import datasets,linear_model 
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score

## Lasso

In [7]:
# 讀取boston房價資料
boston = datasets.load_boston()

# 載入features, 並確認大小
X = boston.data
print("Data shape: ", X.shape)

# 切分訓練集/測試集
x_train, x_test, y_train, y_test = train_test_split(X, boston.target , test_size=0.1, random_state=1)

# 建立一個Lasso回歸模型
lasso = linear_model.Lasso(alpha=0.1)

# 將訓練資料丟進去模型訓練
lasso.fit(x_train, y_train)
print("train score:",lasso.score(x_train, y_train))

# 將測試資料丟進模型得到預測結果
y_pred = lasso.predict(x_test)

Data shape:  (506, 13)
train score: 0.7193621666317072


In [8]:
# 可以看回歸模型的參數值
print('Coefficients: ', lasso.coef_)

# 預測值與實際值的差距，使用 MSE
print("Mean squared error: %.2f"
      % mean_squared_error(y_test, y_pred))

Coefficients:  [-0.10640276  0.05793569 -0.06350259  0.84575491 -0.          3.42213376
 -0.01307927 -1.2623356   0.27745833 -0.0142168  -0.72835262  0.00961902
 -0.56242292]
Mean squared error: 20.98


In [9]:
#確認一下實際值跟預測值
print(y_test[:5])
print(y_pred[:5])

[28.2 23.9 16.6 22.  20.8]
[32.65151645 28.38758412 16.97373178 21.07512569 20.71976419]


HW38使用linear regression時<br>
Mean squared error: 20.54<br>
而這邊使用Lasso<br>
Mean squared error: 20.98<br>
結果並沒有比較準<br>

## Ridge

In [14]:
# 讀取boston房價資料
boston = datasets.load_boston()

# 載入features, 並確認大小
X0 = boston.data
print("Data shape: ", X0.shape)

# 切分訓練集/測試集
x_train0, x_test0, y_train0, y_test0 = train_test_split(X0, boston.target , test_size=0.1, random_state=2)

# 建立一個Lasso回歸模型
ridge = linear_model.Ridge(alpha=0.1)

# 將訓練資料丟進去模型訓練
ridge.fit(x_train0, y_train0)
print("train score:",ridge.score(x_train0, y_train0))

# 將測試資料丟進模型得到預測結果
y_pred0 = ridge.predict(x_test0)

Data shape:  (506, 13)
train score: 0.7434617072440968


In [15]:
# 可以看回歸模型的參數值
print('Coefficients: ', ridge.coef_)

# 預測值與實際值的差距，使用 MSE
print("Mean squared error: %.2f"
      % mean_squared_error(y_test0, y_pred0))

Coefficients:  [-1.11806826e-01  4.33633358e-02 -4.77780647e-03  2.78728265e+00
 -1.41257630e+01  3.81651072e+00 -1.48247875e-03 -1.40719111e+00
  3.25796986e-01 -1.24300847e-02 -9.67766449e-01  1.13760725e-02
 -5.51630945e-01]
Mean squared error: 22.60


In [16]:
#確認一下實際值跟預測值
print(y_test0[:5])
print(y_pred0[:5])

[20.2 15.3 37.3 32.5  8.8]
[23.1507548  20.99199602 34.28097074 31.10030977  3.01932191]


HW38使用linear regression時<br>
Mean squared error: 20.54<br>
Lasso Mean squared error: 20.98<br>
Ridge Mean squared error: 22.60<br>
結果Ridge也沒有比較準, 反而更差<br>
所以如果沒有過擬合發生的話<br>
線性回歸可能會比Lasso或是Ridge好<br>